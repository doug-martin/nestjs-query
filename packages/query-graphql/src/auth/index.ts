export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
