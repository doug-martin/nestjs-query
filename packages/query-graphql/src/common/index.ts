export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
