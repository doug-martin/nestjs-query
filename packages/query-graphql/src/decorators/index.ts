export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
