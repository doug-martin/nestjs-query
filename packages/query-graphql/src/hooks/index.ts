export * from './hooks';
export { createDefaultHook } from './default.hook';
export * from './tokens';
export * from './types';
