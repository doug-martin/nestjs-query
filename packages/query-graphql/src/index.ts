export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
