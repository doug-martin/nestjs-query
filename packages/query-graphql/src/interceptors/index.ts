export * from './hook.interceptor';
export * from './authorizer.interceptor';
