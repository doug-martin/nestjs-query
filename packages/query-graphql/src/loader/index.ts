export * from './find-relations.loader';
export * from './query-relations.loader';
export * from './count-relations.loader';
export * from './aggregate-relations.loader';
export * from './dataloader.factory';
