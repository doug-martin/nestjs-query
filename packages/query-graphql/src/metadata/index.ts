import { GraphQLQueryMetadataStorage } from './metadata-storage';

export { FilterableFieldDescriptor } from './metadata-storage';

/** @internal */
let storage: GraphQLQueryMetadataStorage;
/** @internal */
export function getMetadataStorage(): GraphQLQueryMetadataStorage {
  if (!storage) {
    storage = new GraphQLQueryMetadataStorage();
  }
  return storage;
}
