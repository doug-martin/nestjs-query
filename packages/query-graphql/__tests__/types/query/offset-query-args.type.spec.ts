// eslint-disable-next-line max-classes-per-file
import { SortDirection, SortNulls } from '@nestjs-query/core';
import {
  Args,
  ArgsType,
  Float,
  GraphQLISODateTime,
  GraphQLTimestamp,
  ID,
  Int,
  ObjectType,
  Query,
  Resolver,
} from '@nestjs/graphql';
import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { CursorQueryArgsType, FilterableField, PagingStrategies, QueryArgsType } from '../../../src';
import { generateSchema } from '../../__fixtures__';

describe('Offset paging strategy QueryArgsType with manual options', (): void => {
  afterEach(() => jest.clearAllMocks());

  @ObjectType('TestQuery')
  class TestDto {
    @FilterableField(() => ID)
    idField!: number;

    @FilterableField(() => ID, { nullable: true })
    idFieldOption?: number;

    @FilterableField()
    stringField!: string;

    @FilterableField({ nullable: true })
    stringFieldOptional?: string;

    @FilterableField()
    booleanField!: boolean;

    @FilterableField({ nullable: true })
    booleanFieldOptional?: boolean;

    @FilterableField()
    numberField!: number;

    @FilterableField({ nullable: true })
    numberFieldOptional?: number;

    @FilterableField(() => Float)
    floatField!: number;

    @FilterableField(() => Float, { nullable: true })
    floatFieldOptional?: number;

    @FilterableField(() => Int)
    intField!: number;

    @FilterableField(() => Int, { nullable: true })
    intFieldOptional?: number;

    @FilterableField(() => GraphQLTimestamp)
    timestampField!: Date;

    @FilterableField(() => GraphQLTimestamp, { nullable: true })
    timestampFieldOptional?: Date;

    @FilterableField(() => GraphQLISODateTime)
    date!: Date;

    @FilterableField(() => GraphQLISODateTime, { nullable: true })
    dateOptional?: Date;
  }

  @ObjectType()
  class TestFilterRequiredDto {
    @FilterableField({ filterRequired: true })
    requiredFilterableField!: string;
  }

  @ArgsType()
  class TestOffsetQuery extends QueryArgsType(TestDto, { pagingStrategy: PagingStrategies.OFFSET }) {}

  it('create a query for string fields', async () => {
    @Resolver()
    class TestOffsetQueryArgsResolver {
      @Query(() => String)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args() query: TestOffsetQuery): string {
        return 'hello';
      }
    }
    const schema = await generateSchema([TestOffsetQueryArgsResolver]);
    expect(schema).toMatchSnapshot();
  });

  it('should paging to the correct instance of paging', () => {
    const queryObj: TestOffsetQuery = {
      paging: {
        limit: 10,
        offset: 2,
      },
    };
    const queryInstance = plainToClass(TestOffsetQuery, queryObj);
    expect(validateSync(queryInstance)).toEqual([]);
    expect(queryInstance.paging).toBeInstanceOf(TestOffsetQuery.PageType);
  });

  it('should sorting to the correct instance of sorting', () => {
    const queryObj: TestOffsetQuery = {
      sorting: [{ field: 'stringField', direction: SortDirection.ASC, nulls: SortNulls.NULLS_LAST }],
    };
    const queryInstance = plainToClass(TestOffsetQuery, queryObj);
    expect(validateSync(queryInstance)).toEqual([]);
    expect(queryInstance.sorting![0]).toBeInstanceOf(TestOffsetQuery.SortType);
  });

  it('should make filter to the correct instance of sorting', () => {
    const queryObj: CursorQueryArgsType<TestDto> = {
      filter: {
        stringField: { eq: 'foo' },
      },
    };
    const queryInstance = plainToClass(TestOffsetQuery, queryObj);
    expect(validateSync(queryInstance)).toEqual([]);
    expect(queryInstance.filter).toBeInstanceOf(TestOffsetQuery.FilterType);
  });

  it('should make the filter required if there is a filterRequired field', async () => {
    @ArgsType()
    class TestFilterRequiredQuery extends QueryArgsType(TestFilterRequiredDto, {
      pagingStrategy: PagingStrategies.OFFSET,
    }) {}

    @Resolver()
    class TestOffsetPagingFilterRequiredResolver {
      @Query(() => String)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args() query: TestFilterRequiredQuery): string {
        return 'hello';
      }
    }
    const schema = await generateSchema([TestOffsetPagingFilterRequiredResolver]);
    expect(schema).toMatchSnapshot();
  });

  describe('options', () => {
    @ObjectType()
    class OffsetQueryOptionsDTO extends TestDto {}
    @ArgsType()
    class OffsetQueryOptionsArgs extends QueryArgsType(OffsetQueryOptionsDTO, {
      pagingStrategy: PagingStrategies.OFFSET,
      defaultResultSize: 2,
      maxResultsSize: 5,
      defaultFilter: { booleanField: { is: true } },
      defaultSort: [{ field: 'booleanField', direction: SortDirection.DESC }],
    }) {}

    it('allow apply the options to the generated SDL', async () => {
      @Resolver()
      class TestOffsetQueryManualOptionsResolver {
        @Query(() => String)
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        test(@Args() query: OffsetQueryOptionsArgs): string {
          return 'hello';
        }
      }
      const schema = await generateSchema([TestOffsetQueryManualOptionsResolver]);
      expect(schema).toMatchSnapshot();
    });

    it('should validate a maxResultsSize for paging.limit', () => {
      const queryObj: CursorQueryArgsType<TestDto> = {
        paging: { limit: 10 },
      };
      const QT = QueryArgsType(TestDto, { pagingStrategy: PagingStrategies.OFFSET, maxResultsSize: 5 });
      const queryInstance = plainToClass(QT, queryObj);
      expect(validateSync(queryInstance)).toEqual([
        {
          children: [],
          constraints: {
            PropertyMax: 'Field paging.limit max allowed value is `5`.',
          },
          property: 'paging',
          target: queryObj,
          value: queryObj.paging,
        },
      ]);
    });

    it('should ignore a maxResultsSize for paging.limit if maxResultSize === -1', () => {
      class NoMaxQueryArgsTpe extends QueryArgsType(TestDto, {
        pagingStrategy: PagingStrategies.OFFSET,
        maxResultsSize: -1,
      }) {}
      const queryObj: NoMaxQueryArgsTpe = {
        paging: { limit: 1000 },
      };
      expect(validateSync(plainToClass(NoMaxQueryArgsTpe, queryObj))).toEqual([]);
    });
  });
});
