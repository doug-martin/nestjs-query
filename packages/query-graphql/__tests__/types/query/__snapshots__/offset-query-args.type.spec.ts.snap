// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offset paging strategy QueryArgsType with manual options create a query for string fields 1`] = `
type TestQuery {
  idField: ID!
  idFieldOption: ID
  stringField: String!
  stringFieldOptional: String
  booleanField: Boolean!
  booleanFieldOptional: Boolean
  numberField: Float!
  numberFieldOptional: Float
  floatField: Float!
  floatFieldOptional: Float
  intField: Int!
  intFieldOptional: Int
  timestampField: Timestamp!
  timestampFieldOptional: Timestamp
  date: DateTime!
  dateOptional: DateTime
}

"""
\`Date\` type as integer. Type represents date and time as number of milliseconds from start of UNIX epoch.
"""
scalar Timestamp

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type OffsetQueryOptionsDTO {
  idField: ID!
  idFieldOption: ID
  stringField: String!
  stringFieldOptional: String
  booleanField: Boolean!
  booleanFieldOptional: Boolean
  numberField: Float!
  numberFieldOptional: Float
  floatField: Float!
  floatFieldOptional: Float
  intField: Int!
  intFieldOptional: Int
  timestampField: Timestamp!
  timestampFieldOptional: Timestamp
  date: DateTime!
  dateOptional: DateTime
}

type Query {
  test(
    """Limit or page results."""
    paging: OffsetPaging = {limit: 10}

    """Specify to filter the records returned."""
    filter: TestQueryFilter = {}

    """Specify to sort results."""
    sorting: [TestQuerySort!] = []
  ): String!
}

input OffsetPaging {
  """Limit the number of records returned"""
  limit: Int

  """Offset to start returning records from"""
  offset: Int
}

input TestQueryFilter {
  and: [TestQueryFilter!]
  or: [TestQueryFilter!]
  idField: IDFilterComparison
  idFieldOption: IDFilterComparison
  stringField: StringFieldComparison
  stringFieldOptional: StringFieldComparison
  booleanField: BooleanFieldComparison
  booleanFieldOptional: BooleanFieldComparison
  numberField: NumberFieldComparison
  numberFieldOptional: NumberFieldComparison
  floatField: FloatFieldComparison
  floatFieldOptional: FloatFieldComparison
  intField: IntFieldComparison
  intFieldOptional: IntFieldComparison
  timestampField: TimestampFieldComparison
  timestampFieldOptional: TimestampFieldComparison
  date: DateFieldComparison
  dateOptional: DateFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input BooleanFieldComparison {
  is: Boolean
  isNot: Boolean
}

input NumberFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Float
  neq: Float
  gt: Float
  gte: Float
  lt: Float
  lte: Float
  in: [Float!]
  notIn: [Float!]
  between: NumberFieldComparisonBetween
  notBetween: NumberFieldComparisonBetween
}

input NumberFieldComparisonBetween {
  lower: Float!
  upper: Float!
}

input FloatFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Float
  neq: Float
  gt: Float
  gte: Float
  lt: Float
  lte: Float
  in: [Float!]
  notIn: [Float!]
  between: FloatFieldComparisonBetween
  notBetween: FloatFieldComparisonBetween
}

input FloatFieldComparisonBetween {
  lower: Float!
  upper: Float!
}

input IntFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Int
  neq: Int
  gt: Int
  gte: Int
  lt: Int
  lte: Int
  in: [Int!]
  notIn: [Int!]
  between: IntFieldComparisonBetween
  notBetween: IntFieldComparisonBetween
}

input IntFieldComparisonBetween {
  lower: Int!
  upper: Int!
}

input TimestampFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Timestamp
  neq: Timestamp
  gt: Timestamp
  gte: Timestamp
  lt: Timestamp
  lte: Timestamp
  in: [Timestamp!]
  notIn: [Timestamp!]
  between: TimestampFieldComparisonBetween
  notBetween: TimestampFieldComparisonBetween
}

input TimestampFieldComparisonBetween {
  lower: Timestamp!
  upper: Timestamp!
}

input DateFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: DateTime
  neq: DateTime
  gt: DateTime
  gte: DateTime
  lt: DateTime
  lte: DateTime
  in: [DateTime!]
  notIn: [DateTime!]
  between: DateFieldComparisonBetween
  notBetween: DateFieldComparisonBetween
}

input DateFieldComparisonBetween {
  lower: DateTime!
  upper: DateTime!
}

input TestQuerySort {
  field: TestQuerySortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestQuerySortFields {
  idField
  idFieldOption
  stringField
  stringFieldOptional
  booleanField
  booleanFieldOptional
  numberField
  numberFieldOptional
  floatField
  floatFieldOptional
  intField
  intFieldOptional
  timestampField
  timestampFieldOptional
  date
  dateOptional
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

`;

exports[`Offset paging strategy QueryArgsType with manual options options allow apply the options to the generated SDL 1`] = `
type TestQuery {
  idField: ID!
  idFieldOption: ID
  stringField: String!
  stringFieldOptional: String
  booleanField: Boolean!
  booleanFieldOptional: Boolean
  numberField: Float!
  numberFieldOptional: Float
  floatField: Float!
  floatFieldOptional: Float
  intField: Int!
  intFieldOptional: Int
  timestampField: Timestamp!
  timestampFieldOptional: Timestamp
  date: DateTime!
  dateOptional: DateTime
}

"""
\`Date\` type as integer. Type represents date and time as number of milliseconds from start of UNIX epoch.
"""
scalar Timestamp

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type TestFilterRequiredDto {
  requiredFilterableField: String!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type OffsetQueryOptionsDTO {
  idField: ID!
  idFieldOption: ID
  stringField: String!
  stringFieldOptional: String
  booleanField: Boolean!
  booleanFieldOptional: Boolean
  numberField: Float!
  numberFieldOptional: Float
  floatField: Float!
  floatFieldOptional: Float
  intField: Int!
  intFieldOptional: Int
  timestampField: Timestamp!
  timestampFieldOptional: Timestamp
  date: DateTime!
  dateOptional: DateTime
}

type Query {
  test(
    """Limit or page results."""
    paging: OffsetPaging = {limit: 2}

    """Specify to filter the records returned."""
    filter: OffsetQueryOptionsDTOFilter = {booleanField: {is: true}}

    """Specify to sort results."""
    sorting: [OffsetQueryOptionsDTOSort!] = [{field: booleanField, direction: DESC}]
  ): String!
}

input OffsetPaging {
  """Limit the number of records returned"""
  limit: Int

  """Offset to start returning records from"""
  offset: Int
}

input OffsetQueryOptionsDTOFilter {
  and: [OffsetQueryOptionsDTOFilter!]
  or: [OffsetQueryOptionsDTOFilter!]
  idField: IDFilterComparison
  idFieldOption: IDFilterComparison
  stringField: StringFieldComparison
  stringFieldOptional: StringFieldComparison
  booleanField: BooleanFieldComparison
  booleanFieldOptional: BooleanFieldComparison
  numberField: NumberFieldComparison
  numberFieldOptional: NumberFieldComparison
  floatField: FloatFieldComparison
  floatFieldOptional: FloatFieldComparison
  intField: IntFieldComparison
  intFieldOptional: IntFieldComparison
  timestampField: TimestampFieldComparison
  timestampFieldOptional: TimestampFieldComparison
  date: DateFieldComparison
  dateOptional: DateFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input BooleanFieldComparison {
  is: Boolean
  isNot: Boolean
}

input NumberFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Float
  neq: Float
  gt: Float
  gte: Float
  lt: Float
  lte: Float
  in: [Float!]
  notIn: [Float!]
  between: NumberFieldComparisonBetween
  notBetween: NumberFieldComparisonBetween
}

input NumberFieldComparisonBetween {
  lower: Float!
  upper: Float!
}

input FloatFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Float
  neq: Float
  gt: Float
  gte: Float
  lt: Float
  lte: Float
  in: [Float!]
  notIn: [Float!]
  between: FloatFieldComparisonBetween
  notBetween: FloatFieldComparisonBetween
}

input FloatFieldComparisonBetween {
  lower: Float!
  upper: Float!
}

input IntFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Int
  neq: Int
  gt: Int
  gte: Int
  lt: Int
  lte: Int
  in: [Int!]
  notIn: [Int!]
  between: IntFieldComparisonBetween
  notBetween: IntFieldComparisonBetween
}

input IntFieldComparisonBetween {
  lower: Int!
  upper: Int!
}

input TimestampFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: Timestamp
  neq: Timestamp
  gt: Timestamp
  gte: Timestamp
  lt: Timestamp
  lte: Timestamp
  in: [Timestamp!]
  notIn: [Timestamp!]
  between: TimestampFieldComparisonBetween
  notBetween: TimestampFieldComparisonBetween
}

input TimestampFieldComparisonBetween {
  lower: Timestamp!
  upper: Timestamp!
}

input DateFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: DateTime
  neq: DateTime
  gt: DateTime
  gte: DateTime
  lt: DateTime
  lte: DateTime
  in: [DateTime!]
  notIn: [DateTime!]
  between: DateFieldComparisonBetween
  notBetween: DateFieldComparisonBetween
}

input DateFieldComparisonBetween {
  lower: DateTime!
  upper: DateTime!
}

input OffsetQueryOptionsDTOSort {
  field: OffsetQueryOptionsDTOSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum OffsetQueryOptionsDTOSortFields {
  idField
  idFieldOption
  stringField
  stringFieldOptional
  booleanField
  booleanFieldOptional
  numberField
  numberFieldOptional
  floatField
  floatFieldOptional
  intField
  intFieldOptional
  timestampField
  timestampFieldOptional
  date
  dateOptional
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

`;

exports[`Offset paging strategy QueryArgsType with manual options should make the filter required if there is a filterRequired field 1`] = `
type TestQuery {
  idField: ID!
  idFieldOption: ID
  stringField: String!
  stringFieldOptional: String
  booleanField: Boolean!
  booleanFieldOptional: Boolean
  numberField: Float!
  numberFieldOptional: Float
  floatField: Float!
  floatFieldOptional: Float
  intField: Int!
  intFieldOptional: Int
  timestampField: Timestamp!
  timestampFieldOptional: Timestamp
  date: DateTime!
  dateOptional: DateTime
}

"""
\`Date\` type as integer. Type represents date and time as number of milliseconds from start of UNIX epoch.
"""
scalar Timestamp

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type TestFilterRequiredDto {
  requiredFilterableField: String!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type OffsetQueryOptionsDTO {
  idField: ID!
  idFieldOption: ID
  stringField: String!
  stringFieldOptional: String
  booleanField: Boolean!
  booleanFieldOptional: Boolean
  numberField: Float!
  numberFieldOptional: Float
  floatField: Float!
  floatFieldOptional: Float
  intField: Int!
  intFieldOptional: Int
  timestampField: Timestamp!
  timestampFieldOptional: Timestamp
  date: DateTime!
  dateOptional: DateTime
}

type Query {
  test(
    """Limit or page results."""
    paging: OffsetPaging = {limit: 10}

    """Specify to filter the records returned."""
    filter: TestFilterRequiredDtoFilter!

    """Specify to sort results."""
    sorting: [TestFilterRequiredDtoSort!] = []
  ): String!
}

input OffsetPaging {
  """Limit the number of records returned"""
  limit: Int

  """Offset to start returning records from"""
  offset: Int
}

input TestFilterRequiredDtoFilter {
  and: [TestFilterRequiredDtoFilter!]
  or: [TestFilterRequiredDtoFilter!]
  requiredFilterableField: StringFieldComparison!
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input TestFilterRequiredDtoSort {
  field: TestFilterRequiredDtoSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestFilterRequiredDtoSortFields {
  requiredFilterableField
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

`;
